compiler: improve `for`-loop tuple unpacking

zerbina opened the pull request:
## Summary
- fix an issue with `semForVar` where types were modified, leading to unrelated valid code being rejected by semantic analysis (see `tforloop_unpacking_bug.nim`)
- always use a full copy instead of a shallow copy for `for` variables
- refactor `semForVar`
- refactor `transformYield`
- make `semForVar` `nkError` aware

## Details
- fix the `nkHiddenAddr` nodes produced by `transformYield` having the wrong type (the type of the accessed tuple was used)
- split up `semForVar` into multiple procedures in order to reduce code duplication
- add debug-tracing calls to the `for`-loop related procedures
- improve the documentation of the refactored code
- add tests for `for`-loop tuple unpacking

Unconditionally using `nkFastAsgn` (i.e. a shallow assignment) for local variables is unsound without proving that doing so is safe (which did not happen). Figuring out where to use shallow copies should be left to the cursor inference step.

The `nkFastAsgn` change negatively impacts performance when of the compiled code when using `--gc:refc` (or other legacy GCs), because a full copy is materialized where it previously wasn't. This is not the case for `--gc:arc|orc`, as `nkFastAsgn` was/is treated the same as `nkAsgn` (i.e. as a full copy) for locations with destructors.

---

This PR depends on #487 to be merged first.

zerbina commented:
Hm, I can't seem to reproduce the `tests/realtimeGC/tmain.nim` failure locally.

saem commented:
I'll try in a few, ping me tomorrow please if I forget.

saem commented:
Here is a stacktrace, I had to flip `release` to `debug` for the shared lib compilation on line 24:

```
Traceback (most recent call last)
shared.nim(44)           status
shared.nim(34)           getComm2Status
gc.nim(468)              newObjNoInit
gc_common.nim(325)       prepareDealloc
SIGSEGV: Illegal storage access. (Attempt to read from nil?)
```

saem commented:
Using the above, I disabled all the stacktrace suppression in `gc.nim` and `gc_common.nim` to get the following:

```
Traceback (most recent call last)
shared.nim(44)           status
shared.nim(34)           getComm2Status
gc.nim(468)              newObjNoInit
gc.nim(443)              rawNewObj
gc.nim(827)              collectCT
gc.nim(800)              collectCTBody
gc.nim(761)              collectZCT
gc_common.nim(325)       prepareDealloc
SIGSEGV: Illegal storage access. (Attempt to read from nil?)
```

saem commented:
Stacktrace when running with `-d:useGcAssert` for shared

```
[GCASSERT] incRef: interiorPtrTraceback (most recent call last)
shared.nim(45)           status
strutils.nim(2831)       %
strutils.nim(2774)       addf
gc.nim(284)              unsureAsgnRef
gc.nim(154)              incRef
```

(going to take a closer look)

saem commented:
Grabbed a log from the gc (`-d:logGC`, plus a bunch of fixes for broken debug code) here: https://gist.github.com/saem/6e72621147178a838f5b4b49b6cc639f

relevant snippet:
```
[GC] new cell: 0x7f389dd16040 28 nil rc=0 from shared.nim(14)
[GC] new cell: 0x7f389dd16080 28 nil rc=0 from shared.nim(18) <-- second entry
... spam ...
[GC] zct dealloc cell: 0x7f389dd19040 28 nil rc=0 from strutils.nim(2830)
[GC] zct dealloc cell: 0x7f389dd16400 28 nil rc=0 from shared.nim(45)
[GC] zct dealloc cell: 0x7f389dd163c0 28 nil rc=0 from shared.nim(45)
[GC] zct dealloc cell: 0x7f389dd16380 28 nil rc=0 from shared.nim(45)
[GC] zct dealloc cell: 0x7f389dd16340 28 nil rc=0 from shared.nim(34)
[GC] zct dealloc cell: 0x7f389dd16300 28 nil rc=0 from shared.nim(30)
[GC] zct dealloc cell: 0x7f389dd162c0 28 nil rc=0 from assign.nim(76)
[GC] zct dealloc cell: 0x7f389dd16280 28 nil rc=0 from system.nim(2524)
[GC] zct dealloc cell: 0x7f389dd16240 28 nil rc=0 from assign.nim(76)
[GC] zct dealloc cell: 0x7f389dd16200 28 nil rc=0 from system.nim(2524)
[GC] zct dealloc cell: 0x7f389dd18090 28 nil rc=0 from assign.nim(76)
[GC] zct dealloc cell: 0x7f389dd18040 28 nil rc=0 from system.nim(2524)
[GC] zct dealloc cell: 0x7f389dd17040 28 nil rc=0 from strutils.nim(2830)
[GC] zct dealloc cell: 0x7f389dd161c0 28 nil rc=0 from shared.nim(42)
[GC] zct dealloc cell: 0x7f389dd16180 28 nil rc=0 from shared.nim(42)
[GC] zct dealloc cell: 0x7f389dd16140 28 nil rc=0 from shared.nim(42)
[GC] zct dealloc cell: 0x7f389dd16100 28 nil rc=0 from shared.nim(26)
[GC] zct dealloc cell: 0x7f389dd160c0 28 nil rc=0 from shared.nim(22)
[GC] zct dealloc cell: 0x7f389dd16080 28 nil rc=0 from shared.nim(18)   <--- dealloc'd early
[GC] new cell: 0x7f389dd16040 28 nil rc=0 from system.nim(2524)
[GC] new cell: 0x7f389dd16080 28 nil rc=0 from assign.nim(76)                <--- new cell followed by an inc?!
[GC] incRef: 0x7f389dd16080 28 nil rc=1 from assign.nim(76)
[GCASSERT] incRef: interiorPtr[GCASSERT] statistics:                             <--- assert we know about
.... spam to print gc stats ...
[GC] total memory: 528384
[GC] occupied memory: 61632
[GC] stack scans: 1
[GC] stack cells: 0
[GC] cycle collections: 0
[GC] max threshold: 0
[GC] zct capacity: 1024
[GC] max cycle table size: 0
[GC] max pause time [ms]: 1
[GC] stack bottom: 0x7ffd7e9c6098[GC] max stack size: 0
```

zerbina commented:
I fixed the tests and rebased the PR onto `devel`. CI should succeed now, but I'm leaving this open in case you want to do a final review @saem . Feel free to merge it if you don't encounter any blockers.

zerbina committed:
compiler: improve `for`-loop tuple unpacking

## Summary
- fix an issue with `semForVar` where types were modified, leading to
  unrelated valid code being rejected by semantic analysis (see
   `tforloop_unpacking_bug.nim`)
- always use a full copy instead of a shallow copy for `for` variables
- refactor `semForVar`
- refactor `transformYield`
- make `semForVar` `nkError` aware

## Details
- fix the `nkHiddenAddr` nodes produced by `transformYield` having
  the wrong type (the type of the accessed tuple was used)
- split up `semForVar` into multiple procedures in order to reduce code
  duplication
- add debug-tracing calls to the `for`-loop related procedures
- improve the documentation of the refactored code
- add tests for `for`-loop tuple unpacking

Unconditionally using `nkFastAsgn` (i.e. a shallow assignment) for local
variables is unsound without proving that doing so is safe (which
did not happen). Figuring out where to use shallow copies should be
left to the cursor inference step.

The `nkFastAsgn` change negatively impacts performance of the
compiled code when using `--gc:refc` (or other legacy GCs), because a
full copy is materialized where it previously wasn't. This is not the
case for `--gc:arc|orc`, as `nkFastAsgn` was/is treated the same as
`nkAsgn` (i.e. as a full copy) for locations with destructors.

saem commented:
Bors r+

Yay.

saem commented:
Copying @zerbina's comments from chat with respect to debugging the real-time GC test failure:

> Heh, it turned out to be a rather simple cause in the end: a dynamic library is used, but without using `-d:useNimRtl`. In other words, there exist two instances of the GC.
> The GC instance being used by `shared.nim` has a stack bottom that points a few hundred bytes past the actual stack beginning (due to `setStackBottom` being called from inside the dynlib's entry point). This causes the stack scanning to not detect the `result` variable of `strutils.%` (because it is located at a stack address that comes _before_ what the GC considers to be the stack bottom), thus leading to the string being collected.
> Compiling both `tmain.nim` and `shared.nim` with `-d:useNimRtl` fixes the crash.

Link: https://matrix.to/#/!MAXxvfCYnfYgZqSizw:matrix.org/$lP8QF1MTEnbKCOozooiiJyOCd6OtOwGgkHbcZOL6r68?via=matrix.org&via=libera.chat&via=envs.net
